File: androlua/src/main/java/cn/vove7/androlua/luautils/LuaEditor.java
package cn.vove7.androlua.luautils;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.ActionMode;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.RadioGroup.LayoutParams;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;
import android.widget.Toast;

import com.myopicmobile.textwarrior.android.FreeScrollingTextField;
import com.myopicmobile.textwarrior.android.YoyoNavigationMethod;
import com.myopicmobile.textwarrior.common.ColorScheme;
import com.myopicmobile.textwarrior.common.ColorSchemeDark;
import com.myopicmobile.textwarrior.common.ColorSchemeLight;
import com.myopicmobile.textwarrior.common.Document;
import com.myopicmobile.textwarrior.common.DocumentProvider;
import com.myopicmobile.textwarrior.common.LanguageLua;
import com.myopicmobile.textwarrior.common.Lexer;
import com.myopicmobile.textwarrior.common.LinearSearchStrategy;
import com.myopicmobile.textwarrior.common.ReadTask;

import java.io.File;

import cn.vove7.vtp.toast.Voast;

public class LuaEditor extends FreeScrollingTextField {

    private Document _inputingDoc;

    private boolean _isWordWrap;

    private Context mContext;

    private String _lastSelectedFile;

    private String fontDir = "/sdcard/fonts/";

    private int _index;
    private LinearSearchStrategy finder;
    private int idx;
    private String mKeyword;

    private void initFont() {
        File df = new File(fontDir + "default.ttf");
        if (df.exists())
            setTypeface(Typeface.createFromFile(df));
        File bf = new File(fontDir + "bold.ttf");
        if (bf.exists())
            setBoldTypeface(Typeface.createFromFile(bf));
        File tf = new File(fontDir + "italic.ttf");
        if (tf.exists())
            setItalicTypeface(Typeface.createFromFile(tf));
    }

    public LuaEditor(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
        setTypeface(Typeface.MONOSPACE);
        DisplayMetrics dm = context.getResources().getDisplayMetrics();
        initFont();
        float size = TypedValue.applyDimension(2, BASE_TEXT_SIZE_PIXELS, dm);
        setTextSize((int) size);
        setShowLineNumbers(true);
        setHighlightCurrentRow(true);
        setWordWrap(false);
        setAutoIndentWidth(2);
        Lexer.setLanguage(LanguageLua.getInstance());
        setNavigationMethod(new YoyoNavigationMethod(this));
        TypedArray array = mContext.getTheme().obtainStyledAttributes(new int[]{
                android.R.attr.colorBackground,
                android.R.attr.textColorPrimary,
                android.R.attr.textColorHighlight,
        });
        int backgroundColor = array.getColor(0, 0xFF00FF);
        int textColor = array.getColor(1, 0xFF00FF);
        int textColorHighlight = array.getColor(2, 0xFF00FF);
        array.recycle();
        setTextColor(textColor);
        setTextHighligtColor(textColorHighlight);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {

        super.onLayout(changed, left, top, right, bottom);
        if (_index != 0 && right > 0) {
            moveCaret(_index);
            _index = 0;
        }
    }

    public void setDark(boolean isDark) {
        if (isDark)
            setColorScheme(new ColorSchemeDark());
        else
            setColorScheme(new ColorSchemeLight());
    }

    public void addNames(String[] names) {
        LanguageLua lang = (LanguageLua) Lexer.getLanguage();
        String[] old = lang.getNames();
        String[] news = new String[old.length + names.length];
        System.arraycopy(old, 0, news, 0, old.length);
        System.arraycopy(names, 0, news, old.length, names.length);
        lang.setNames(news);
        Lexer.setLanguage(lang);
        respan();
        invalidate();

    }

    public void setPanelBackgroundColor(int color) {

        _autoCompletePanel.setBackgroundColor(color);
    }

    public void setPanelTextColor(int color) {

        _autoCompletePanel.setTextColor(color);
    }

    public void setKeywordColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.KEYWORD, color);
    }

    public void setUserwordColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.LITERAL, color);
    }

    public void setBasewordColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.NAME, color);
    }

    public void setStringColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.STRING, color);
    }

    public void setCommentColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.COMMENT, color);
    }

    public void setBackgoudColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.BACKGROUND, color);
    }

    public void setTextColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.FOREGROUND, color);
    }

    public void setTextHighligtColor(int color) {
        getColorScheme().setColor(ColorScheme.Colorable.SELECTION_BACKGROUND, color);
    }

    public String getSelectedText() {

        return _hDoc.subSequence(getSelectionStart(), getSelectionEnd() - getSelectionStart()).toString();
    }

    @Override
    public boolean onKeyShortcut(int keyCode, KeyEvent event) {
        final int filteredMetaState = event.getMetaState() & ~KeyEvent.META_CTRL_MASK;
        if (KeyEvent.metaStateHasNoModifiers(filteredMetaState)) {
            switch (keyCode) {
                case KeyEvent.KEYCODE_A:
                    selectAll();
                    return true;
                case KeyEvent.KEYCODE_X:
                    cut();
                    return true;
                case KeyEvent.KEYCODE_C:
                    copy();
                    return true;
                case KeyEvent.KEYCODE_V:
                    paste();
                    return true;
            }
        }
        return super.onKeyShortcut(keyCode, event);
    }

    public void gotoLine() {
        startGotoMode();
    }

    public void search() {
        startFindMode();
    }

    public void startGotoMode() {

        startActionMode(new ActionMode.Callback() {
            private int idx;
            private EditText edit;

            @Override
            public boolean onCreateActionMode(ActionMode mode, Menu menu) {

                mode.setTitle("转到");
                mode.setSubtitle(null);

                edit = new android.support.v7.widget.AppCompatEditText(mContext) {
                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {
                        if (s.length() > 0) {
                            idx = 0;
                            _gotoLine();
                        }
                    }

                };

                edit.setSingleLine(true);
                edit.setInputType(2);
                edit.setImeOptions(2);
                edit.setOnEditorActionListener(new OnEditorActionListener() {

                    @Override
                    public boolean onEditorAction(TextView p1, int p2, KeyEvent p3) {

                        _gotoLine();
                        return true;
                    }
                });
                edit.setLayoutParams(new LayoutParams(getWidth() / 3, -1));
                menu.add(0, 1, 0, "").setActionView(edit);
                menu.add(0, 2, 0, mContext.getString(android.R.string.ok));
                edit.requestFocus();
                return true;
            }

            private void _gotoLine() {
                String s = edit.getText().toString();
                if (s.isEmpty())
                    return;

                int l = Integer.valueOf(s).intValue();
                if (l > _hDoc.getRowCount()) {
                    l = _hDoc.getRowCount();
                }
                gotoLine(l);

            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {

                return false;
            }

            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {

                switch (item.getItemId()) {
                    case 1:
                        break;
                    case 2:
                        _gotoLine();
                        break;

                }
                return false;
            }

            @Override
            public void onDestroyActionMode(ActionMode p1) {

            }
        });

    }

    public void startFindMode() {

        startActionMode(new ActionMode.Callback() {

            private LinearSearchStrategy finder;

            private int idx;

            private EditText edit;

            @Override
            public boolean onCreateActionMode(ActionMode mode, Menu menu) {

                mode.setTitle("搜索");
                mode.setSubtitle(null);

                edit = new android.support.v7.widget.AppCompatEditText(mContext) {
                    @Override
                    public void onTextChanged(CharSequence s, int start, int before, int count) {
                        if (s.length() > 0) {
                            idx = 0;
                            findNext();
                        }
                    }
                };
                edit.setSingleLine(true);
                edit.setImeOptions(3);
                edit.setOnEditorActionListener(new OnEditorActionListener() {

                    @Override
                    public boolean onEditorAction(TextView p1, int p2, KeyEvent p3) {

                        findNext();
                        return true;
                    }
                });
                edit.setLayoutParams(new LayoutParams(getWidth() / 3, -1));
                menu.add(0, 1, 0, "").setActionView(edit);
                menu.add(0, 2, 0, mContext.getString(android.R.string.search_go));
                edit.requestFocus();
                return true;
            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
                return false;
            }

            @Override
            public boolean onActionItemClicked(ActionMode mode, MenuItem item) {

                switch (item.getItemId()) {
                    case 1:
                        break;
                    case 2:
                        findNext();
                        break;

                }
                return false;
            }

            private void findNext() {
                finder = new LinearSearchStrategy();
                String kw = edit.getText().toString();
                if (kw.isEmpty()) {
                    selectText(false);
                    return;
                }
                idx = finder.find(getText(), kw, idx, getText().length(), false, false);
                if (idx == -1) {
                    selectText(false);
                    Toast.makeText(mContext, "未找到", Toast.LENGTH_SHORT).show();
                    idx = 0;
                    return;
                }
                setSelection(idx, edit.getText().length());
                idx += edit.getText().length();
                moveCaret(idx);
            }

            @Override
            public void onDestroyActionMode(ActionMode p1) {
            }
        });

    }

    @Override
    public void setWordWrap(boolean enable) {
        _isWordWrap = enable;
        super.setWordWrap(enable);
    }

    public DocumentProvider getText() {
        return createDocumentProvider();
    }

    @Override
    public void setText(CharSequence c) {
        //TextBuffer text=new TextBuffer();
        Document doc = new Document(this);
        doc.setWordWrap(_isWordWrap);
        doc.setText(c);
        setDocumentProvider(new DocumentProvider(doc));
        //doc.analyzeWordWrap();
    }

    public void insert(int idx, String text) {
        selectText(false);
        moveCaret(idx);
        paste(text);
        //_hDoc.insert(idx,text);
    }

    public void setText(CharSequence c, boolean isRep) {
        replaceText(0, getLength() - 1, c.toString());
    }

    public void setSelection(int index) {
        selectText(false);
        if (!hasLayout())
            moveCaret(index);
        else
            _index = index;
    }

    public void gotoLine(int line) {
        if (line > _hDoc.getRowCount()) {
            line = _hDoc.getRowCount();
        }
        int i = getText().getLineOffset(line - 1);
        setSelection(i);
    }

    public void undo() {
        DocumentProvider doc = createDocumentProvider();
        int newPosition = doc.undo();

        if (newPosition >= 0) {
            //TODO editor.setEdited(false); if reached original condition of file
            setEdited(true);

            respan();
            selectText(false);
            moveCaret(newPosition);
            invalidate();
        }

    }

    public void redo() {
        DocumentProvider doc = createDocumentProvider();
        int newPosition = doc.redo();

        if (newPosition >= 0) {
            setEdited(true);

            respan();
            selectText(false);
            moveCaret(newPosition);
            invalidate();
        }

    }

    public void open(String filename) {
        _lastSelectedFile = filename;

        File inputFile = new File(filename);
        _inputingDoc = new Document(this);
        _inputingDoc.setWordWrap(this.isWordWrap());
        ReadTask _taskRead = new ReadTask(this, inputFile);
        _taskRead.start();
    }

    public void save(String filename, boolean overwrite) {
        File outputFile = new File(filename);

        if (outputFile.exists()) {
            if (!outputFile.canWrite()) {
                return;
            }
        }

    }

    public boolean findNext(String keyword) {
        if (!keyword.equals(mKeyword)) {
            mKeyword = keyword;
            idx = 0;
        }

        finder = new LinearSearchStrategy();
        String kw = mKeyword;
        if (kw.isEmpty()) {
            selectText(false);
            return false;
        }
        idx = finder.find(getText(), kw, idx, getText().length(), false, false);
        if (idx == -1) {
            selectText(false);
            Toast.makeText(mContext, "未找到", Toast.LENGTH_SHORT).show();
            idx = 0;
            return false;
        }
        setSelection(idx, mKeyword.length());
        idx += mKeyword.length();
        moveCaret(idx);
        return true;
    }

}

File: androlua/src/main/java/cn/vove7/androlua/luautils/LuaEditor.kt
package cn.vove7.androlua.luautils

import android.content.Context
import android.graphics.Typeface
import android.util.AttributeSet
import android.util.TypedValue
import android.view.ActionMode
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import android.widget.RadioGroup.LayoutParams
import android.widget.Toast
import cn.vove7.common.interfaces.CodeEditorOperation
import com.myopicmobile.textwarrior.android.FreeScrollingTextField
import com.myopicmobile.textwarrior.android.YoyoNavigationMethod
import com.myopicmobile.textwarrior.common.*
import com.myopicmobile.textwarrior.lua.LanguageLua

class LuaEditor(private val mContext: Context, attrs: AttributeSet)
    : FreeScrollingTextField(mContext, attrs), CodeEditorOperation {

    private val _inputingDoc: Document? = null

    private var _isWordWrap: Boolean = false

    private val _lastSelectedFile: String? = null

    //private String fontDir = "/sdcard/fonts/";

    private var _index: Int = 0
    private var finder: LinearSearchStrategy? = null
    private var idx: Int = 0
    private var mKeyword: String? = null

    override val selectedText: String?
        get() = _hDoc.subSequence(selectionStart, selectionEnd - selectionStart) as String
    override var isEdit: Boolean = false
        get() = isEdited

    val text: DocumentProvider
        get() = createDocumentProvider()

    init {
        setTypeface(Typeface.MONOSPACE)
        val dm = mContext.resources.displayMetrics
        //initFont();
        val size = TypedValue.applyDimension(2, FreeScrollingTextField.BASE_TEXT_SIZE_PIXELS.toFloat(), dm)
        setTextSize(size.toInt())
        isShowLineNumbers = true
        setHighlightCurrentRow(true)
        isWordWrap = false
        autoIndentWidth = 2
        Lexer.setLanguage(LanguageLua.instance)
        setNavigationMethod(YoyoNavigationMethod(this))
        val array = mContext.theme.obtainStyledAttributes(intArrayOf(android.R.attr.colorBackground, android.R.attr.textColorPrimary, android.R.attr.textColorHighlight))
        val backgroundColor = array.getColor(0, 0xFF00FF)
        val textColor = array.getColor(1, 0xFF00FF)
        val textColorHighlight = array.getColor(2, 0xFF00FF)
        array.recycle()
        setTextColor(textColor)
        setTextHighligtColor(textColorHighlight)
//        setDark(true)
        colorScheme = ColorSchemeDark()
    }


    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        if (_index != 0 && right > 0) {
            moveCaret(_index)
            _index = 0
        }
    }

    override fun getEditorContent(): String {
        return text.toString()
    }

    override fun setDark(isDark: Boolean) {
        if (isDark)
            colorScheme = ColorSchemeDark()
        else
            colorScheme = ColorSchemeLight()
    }

    fun addNames(names: Array<String>) {
        val lang = Lexer.getLanguage() as LanguageLua
        val old = lang.names
        val news = arrayOfNulls<String>(old.size + names.size)
        System.arraycopy(old, 0, news, 0, old.size)
        System.arraycopy(names, 0, news, old.size, names.size)
        lang.names = news
        Lexer.setLanguage(lang)
        respan()
        invalidate()

    }

    fun setPanelBackgroundColor(color: Int) {

        _autoCompletePanel.setBackgroundColor(color)
    }

    fun setPanelTextColor(color: Int) {

        _autoCompletePanel.setTextColor(color)
    }

    fun setKeywordColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.KEYWORD, color)
    }

    fun setUserwordColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.LITERAL, color)
    }

    fun setBasewordColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.NAME, color)
    }

    fun setStringColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.STRING, color)
    }

    fun setCommentColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.COMMENT, color)
    }

    fun setBackgoudColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.BACKGROUND, color)
    }

    fun setTextColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.FOREGROUND, color)
    }

    fun setTextHighligtColor(color: Int) {
        colorScheme.setColor(ColorScheme.Colorable.SELECTION_BACKGROUND, color)
    }

    override fun onKeyShortcut(keyCode: Int, event: KeyEvent): Boolean {
        val filteredMetaState = event.metaState and KeyEvent.META_CTRL_MASK.inv()
        if (KeyEvent.metaStateHasNoModifiers(filteredMetaState)) {
            when (keyCode) {
                KeyEvent.KEYCODE_A -> {
                    selectAll()
                    return true
                }
                KeyEvent.KEYCODE_X -> {
                    cut()
                    return true
                }
                KeyEvent.KEYCODE_C -> {
                    copy()
                    return true
                }
                KeyEvent.KEYCODE_V -> {
                    paste()
                    return true
                }
            }
        }
        return super.onKeyShortcut(keyCode, event)
    }

    override fun gotoLine() {
        startGotoMode()
    }

    override fun insert(s: CharSequence?) {
        insert(selectionStart, s)
    }

    override fun find() {
        startFindMode()
    }

    fun search() {
        startFindMode()
    }

    fun startGotoMode() {

        startActionMode(object : ActionMode.Callback {
            private var idx: Int = 0
            private var edit: EditText? = null

            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                mode.title = "跳转行"
                mode.subtitle = null
                edit = object : androidx.appcompat.widget.AppCompatEditText(mContext) {
                    public override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                        if (s.length > 0) {
                            idx = 0
                            _gotoLine()
                        }
                    }
                }
                edit!!.setSingleLine(true)
                edit!!.inputType = 2
                edit!!.imeOptions = 2
                edit!!.setOnEditorActionListener { p1, p2, p3 ->
                    _gotoLine()
                    true
                }
                edit!!.layoutParams = LayoutParams(width / 3, -1)
                menu.add(0, 1, 0, "").actionView = edit
                menu.add(0, 2, 0, mContext.getString(android.R.string.ok))
                edit!!.requestFocus()
                return true
            }

            private fun _gotoLine() {
                val s = edit!!.text.toString()
                if (s.isEmpty())
                    return
                var l = Integer.valueOf(s)
                if (l > _hDoc.rowCount) {
                    l = _hDoc.rowCount
                }
                gotoLine(l)
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                when (item.itemId) {
                    1 -> {
                    }
                    2 -> _gotoLine()
                }
                return false
            }

            override fun onDestroyActionMode(p1: ActionMode) {}
        })
    }

    fun startFindMode() {
        startActionMode(object : ActionMode.Callback {
            private var finder: LinearSearchStrategy? = null
            private var idx: Int = 0
            private var edit: EditText? = null

            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                mode.title = "查找"
                mode.subtitle = null
                edit = object : androidx.appcompat.widget.AppCompatEditText(mContext) {
                    public override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                        if (s.length > 0) {
                            idx = 0
                            findNext()
                        }
                    }
                }
                edit!!.setSingleLine(true)
                edit!!.imeOptions = 3
                edit!!.setOnEditorActionListener { p1, p2, p3 ->
                    findNext()
                    true
                }
                edit!!.layoutParams = LayoutParams(width / 3, -1)
                menu.add(0, 1, 0, "").actionView = edit
                menu.add(0, 2, 0, mContext.getString(android.R.string.search_go))
                edit!!.requestFocus()
                return true
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {

                when (item.itemId) {
                    1 -> {
                    }
                    2 -> findNext()
                }
                return false
            }

            private fun findNext() {
                finder = LinearSearchStrategy()
                val kw = edit!!.text.toString()
                if (kw.isEmpty()) {
                    selectText(false)
                    return
                }
                idx = finder!!.find(text, kw, idx, text.length, false, false)
                if (idx == -1) {
                    selectText(false)
                    Toast.makeText(mContext, "未找到", Toast.LENGTH_SHORT).show()
                    idx = 0
                    return
                }
                setSelection(idx, edit!!.text.length)
                idx += edit!!.text.length
                moveCaret(idx)
            }

            override fun onDestroyActionMode(p1: ActionMode) {}
        })

    }

    override fun setWordWrap(enable: Boolean) {
        _isWordWrap = enable
        super.setWordWrap(enable)
    }

    override fun setText(c: CharSequence) {
        //TextBuffer text=new TextBuffer();
        val doc = Document(this)
        doc.isWordWrap = _isWordWrap
        doc.setText(c)
        setDocumentProvider(DocumentProvider(doc))
        //doc.analyzeWordWrap();
    }

    fun insert(idx: Int, text: CharSequence?) {
        selectText(false)
        moveCaret(idx)
        paste(text?.toString() ?: "")
        //_hDoc.insert(idx,text);
    }

    fun setText(c: CharSequence, isRep: Boolean) {
        replaceText(0, length - 1, c.toString())
    }

    fun setSelection(index: Int) {
        selectText(false)
        if (!hasLayout())
            moveCaret(index)
        else
            _index = index
    }

    fun gotoLine(line: Int) {
        var line = line
        if (line > _hDoc.rowCount) {
            line = _hDoc.rowCount
        }
        val i = text.getLineOffset(line - 1)
        setSelection(i)
    }

    override fun undo() {
        val doc = createDocumentProvider()
        val newPosition = doc.undo()

        if (newPosition >= 0) {
            //TODO editor.setEdited(false); if reached original condition of file
            isEdited = true

            respan()
            selectText(false)
            moveCaret(newPosition)
            invalidate()
        }

    }

    override fun redo() {
        val doc = createDocumentProvider()
        val newPosition = doc.redo()

        if (newPosition >= 0) {
            isEdited = true
            respan()
            selectText(false)
            moveCaret(newPosition)
            invalidate()
        }

    }
    //
    //@Override
    //public void openFile(@NonNull String filename) {
    //    //_lastSelectedFile = filename;
    //    //
    //    //File inputFile = new File(filename);
    //    //_inputingDoc = new Document(this);
    //    //_inputingDoc.setWordWrap(this.isWordWrap());
    //    //ReadTask _taskRead = new ReadTask(this, inputFile);
    //    //_taskRead.start();
    //}


    fun findNext(keyword: String): Boolean {
        if (keyword != mKeyword) {
            mKeyword = keyword
            idx = 0
        }

        finder = LinearSearchStrategy()
        val kw = mKeyword
        if (kw!!.isEmpty()) {
            selectText(false)
            return false
        }
        idx = finder!!.find(text, kw, idx, text.length, false, false)
        if (idx == -1) {
            selectText(false)
            Toast.makeText(mContext, "未找到", Toast.LENGTH_SHORT).show()
            idx = 0
            return false
        }
        setSelection(idx, mKeyword!!.length)
        idx += mKeyword!!.length
        moveCaret(idx)
        return true
    }

}
